assets/js/base/context/cart-checkout/checkout/processor/index.js

```diff
@@ -39,10 +43,28 @@ const CheckoutProcessor = () => {
 	const currentBillingData = useRef( billingData );
 	const currentShippingAddress = useRef( shippingAddress );
 
-	const withErrors = hasValidationErrors();
+	const checkoutWillHaveError =
+		hasValidationErrors ||
+		currentPaymentStatus.hasError ||
+		shippingHasError;
+
+	useEffect( () => {
+		if (
+			checkoutWillHaveError !== checkoutHasError &&
+			( checkoutIsProcessing || checkoutIsProcessingComplete )
+		) {
+			dispatchActions.setHasError( checkoutWillHaveError );
+		}
+	}, [
+		checkoutWillHaveError,
+		checkoutHasError,
+		checkoutIsProcessing,
+		checkoutIsProcessingComplete,
+	] );
```

Comment:
Instead of every consumer dispatching `setHasError` to update Checkout's error state, the Checkout processor subscribes to the consumers and dispatches error state updates internally. In my opinion it simplifies the logic because everything is more self-contained. I think this change makes #2096 irrelevant now, but we can reopen it if we still think it would be good for extensions to have a finer control on the number of errors in the Checkout state.